protocol/metrics/consumer_optimizer_qos_client.go

```diff
@@ -0,0 +1,275 @@
+	// must be called under read lock
+	return coqc.getProviderChainMapCounterValue(coqc.chainIdToProviderToRelaysCount, chainId, providerAddress)
+}
+
+func (coqc *ConsumerOptimizerQoSClient) getProviderChainNodeErrorsCount(chainId, providerAddress string) uint64 {
+	// must be called under read lock
+	return coqc.getProviderChainMapCounterValue(coqc.chainIdToProviderToNodeErrorsCount, chainId, providerAddress)
+}
+
+func (coqc *ConsumerOptimizerQoSClient) getProviderChainStake(chainId, providerAddress string, epoch uint64) int64 {
+	// must be called under read lock
+	if providersMap, found := coqc.chainIdToProviderToEpochToStake[chainId]; found {
+		if epochMap, found := providersMap[providerAddress]; found {
+			if stake, found := epochMap[epoch]; found {
+				return stake
+			}
+		}
+	}
+	return 0
+}
+
+func (coqc *ConsumerOptimizerQoSClient) calculateNodeErrorRate(chainId, providerAddress string) float64 {
+	// must be called under read lock
+	relaysCount := coqc.getProviderChainRelaysCount(chainId, providerAddress)
+	if relaysCount > 0 {
+		errorsCount := coqc.getProviderChainNodeErrorsCount(chainId, providerAddress)
+		return float64(errorsCount) / float64(relaysCount)
+	}
+
+	return 0
+}
+
+func (coqc *ConsumerOptimizerQoSClient) appendOptimizerQoSReport(report OptimizerQoSReport, chainId string, epoch uint64) {
+	// must be called under read lock
+	if coqc == nil {
+		return
+	}
+
+	optimizerQoSReportToSend := optimizerQoSReportToSend{
+		Timestamp:         time.Now(),
+		ConsumerOrigin:    coqc.consumerOrigin,
+		SyncScore:         report.SyncScore,
+		AvailabilityScore: report.AvailabilityScore,
+		LatencyScore:      report.LatencyScore,
+		GenericScore:      report.GenericScore,
+		ProviderAddress:   report.ProviderAddress,
+		ChainId:           chainId,
+		Epoch:             epoch,
+		NodeErrorRate:     coqc.calculateNodeErrorRate(chainId, report.ProviderAddress),
+		ProviderStake:     coqc.getProviderChainStake(chainId, report.ProviderAddress, epoch),
+	}
+
+	coqc.queueSender.appendQueue(optimizerQoSReportToSend)
+}
+
+func (coqc *ConsumerOptimizerQoSClient) getReportsFromOptimizers() {
+	coqc.lock.RLock() // we only read from the maps here
+	defer coqc.lock.RUnlock()
+
+	ignoredProviders := map[string]struct{}{}
+	cu := uint64(10)
+	requestedBlock := spectypes.LATEST_BLOCK
+
+	currentEpoch := atomic.LoadUint64(&coqc.atomicCurrentEpoch)
+
+	for chainId, optimizer := range coqc.optimizers {
+		providersMap, ok := coqc.chainIdToProviderToEpochToStake[chainId]
+		if !ok {
+			continue
+		}
+
+		reports := optimizer.CalculateQoSScoresForMetrics(maps.Keys(providersMap), ignoredProviders, cu, requestedBlock)
+		for _, report := range reports {
+			coqc.appendOptimizerQoSReport(report, chainId, currentEpoch)
+		}
+	}
+}
+
+func (coqc *ConsumerOptimizerQoSClient) StartOptimizersQoSReportsCollecting(ctx context.Context, samplingInterval time.Duration) {
+	utils.LavaFormatTrace("Starting ConsumerOptimizerQoSClient reports collecting")
```

Comment:
let's align to check for nil in all public functions

Reply:
Moved in 26009a9eb8f0e16e3dcd419d9b4314d3fe1bbd24